File: portal/PortalSvr/RabbitMqReceivers/RabbitMqReceiver.cs
/*
* Copyright (c) 2021 PSPACE, inc. KSAN Development Team ksan@pspace.co.kr
* KSAN is a suite of free software: you can redistribute it and/or modify it under the terms of
* the GNU General Public License as published by the Free Software Foundation, either version
* 3 of the License.  See LICENSE for details
*
* 본 프로그램 및 관련 소스코드, 문서 등 모든 자료는 있는 그대로 제공이 됩니다.
* KSAN 프로젝트의 개발자 및 개발사는 이 프로그램을 사용한 결과에 따른 어떠한 책임도 지지 않습니다.
* KSAN 개발팀은 사전 공지, 허락, 동의 없이 KSAN 개발에 관련된 모든 결과물에 대한 LICENSE 방식을 변경 할 권리가 있습니다.
*/
using System;
using System.Threading;
using System.Threading.Tasks;
using PortalData;
using PortalProvider.Providers.RabbitMq;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using MTLib.Core;
using MTLib.Reflection;
using Newtonsoft.Json;
using RabbitMQ.Client;
using RabbitMQ.Client.Events;

// ReSharper disable TemplateIsNotCompileTimeConstantProblem

namespace PortalSvr.RabbitMqReceivers
{
	/// <summary>Rabbit MQ로 부터 정보를 수신하는 클래스</summary>
	public abstract class RabbitMqReceiver : BackgroundService
	{
		/// <summary>Rabbit MQ 연결 객체</summary>
		private IConnection m_connection;
		/// <summary>Rabbit MQ 채널 객체</summary>
		private IModel m_channel;
		/// <summary>큐 이름</summary>
		private readonly string m_queueName;
		/// <summary>Exchange 이름</summary>
		private readonly string m_exchangeName;
		/// <summary>바인딩 키 목록</summary>
		private readonly string[] m_bindingKeys;
		/// <summary>Rabbit MQ 설정 객체</summary>
		private readonly RabbitMqConfiguration m_config = new RabbitMqConfiguration();
		/// <summary>로거</summary>
		protected readonly ILogger m_logger;
		/// <summary>서비스 팩토리</summary>
		protected readonly IServiceScopeFactory m_serviceScopeFactory;

		/// <summary>생성자</summary>
		/// <param name="queueName">큐 이름</param>
		/// <param name="exchangeName">Exchange 이름</param>
		/// <param name="bindingKeys">바인딩 키 목록</param>
		/// <param name="rabbitMqOptions">Rabbit MQ 설정 옵션 객체</param>
		/// <param name="logger">로거</param>
		/// <param name="serviceScopeFactory">서비스 팩토리</param>
		public RabbitMqReceiver(
			string queueName,
			string exchangeName,
			string[] bindingKeys,
			IOptions<RabbitMqConfiguration> rabbitMqOptions,
			ILogger logger,
			IServiceScopeFactory serviceScopeFactory
		)
		{
			try
			{
				// 큐 이름 저장
				m_queueName = queueName;
				// Exchange 이름 저장
				m_exchangeName = exchangeName;
				// 바인딩 키 목록 저장
				m_bindingKeys = bindingKeys;
				// 설정 복사
				m_config.CopyValueFrom(rabbitMqOptions.Value);
				// 로거
				m_logger = logger;
				// 서비스 팩토리 저장
				m_serviceScopeFactory = serviceScopeFactory;
				// Rabbit MQ 리스너 초기화
				InitializeRabbitMqListener();
			}
			catch (Exception ex)
			{
				NNException.Log(ex);
				throw;
			}
		}

		/// <summary>Rabbit MQ 리스너 초기화</summary>
		private void InitializeRabbitMqListener()
		{
			try
			{
				var Factory = new ConnectionFactory
				{
					ClientProvidedName = m_config.Name,
					HostName = m_config.Host,
					Port = m_config.Port,
					VirtualHost = m_config.VirtualHost,
					UserName = m_config.User,
					Password = m_config.Password
				};

				// 연결 객체 생성
				m_connection = Factory.CreateConnection();
				// 연결 해제 이벤트 설정
				m_connection.ConnectionShutdown += (_, _) =>
				{
					m_logger.LogInformation($"[Rabbit MQ] Connection is shutdown. {m_queueName}");
				};
				// 채널 생성
				m_channel = m_connection.CreateModel();
				// 큐 설정
				m_channel.QueueDeclare(queue: m_queueName, durable: true, exclusive: false, autoDelete: false, arguments: null);
				// Exchange 설절
				m_channel.ExchangeDeclare(exchange: m_exchangeName, type: "topic");
				// 모든 바인딩 키 처리
				if (m_bindingKeys != null)
				{
					foreach (string bindingKey in m_bindingKeys)
						m_channel.QueueBind(queue: m_queueName, exchange: m_exchangeName, routingKey: bindingKey);
				}
			}
			catch (Exception ex)
			{
				NNException.Log(ex);
			}
		}

		/// <summary>서비스 실행</summary>
		/// <param name="StoppingToken">실행 취소 토큰 객체</param>
		/// <returns></returns>
		protected override Task ExecuteAsync(CancellationToken StoppingToken)
		{
			try
			{
				StoppingToken.ThrowIfCancellationRequested();

				// Consumer 생성
				var Consumer = new EventingBasicConsumer(m_channel);
				// 데이터 수신 이벤트 설정
				Consumer.Received += async (_, ea) =>
				{
					try
					{
						// 수신 객체가 유효한 경우, 해당 데이터 처리
						if (ea != null)
						{
							var ResponseHandleMessage = await HandleMessage(ea.RoutingKey, ea.Body.ToArray());

							// 처리된 건인 경우
							if (ResponseHandleMessage.IsProcessed)
							{
								m_logger.LogDebug($"[Process] MQ Message Received on [{{0}}] : {{1}}, Data = {{2}}", m_queueName, ea.RoutingKey, ea.Body.ToArray().GetString());

								// 응답 연관 아이디가 존재하는 경우
								var Properties = ea.BasicProperties;
								if (Properties != null && !Properties.CorrelationId.IsEmpty())
								{
									// 답으로 보낼 속성 생성하고 연관 아이디 저장
									var ReplyProps = m_channel.CreateBasicProperties();
									ReplyProps.CorrelationId = Properties.CorrelationId;

									// 결과를 json으로 변환
									string json = JsonConvert.SerializeObject(ResponseHandleMessage);

									// 응답 전송
									m_channel.BasicPublish(exchange: "",
										routingKey: Properties.ReplyTo,
										basicProperties: ReplyProps,
										body: json.GetBytes());
								}

								// 승인 처리
								m_channel.BasicAck(ea.DeliveryTag, false);
							}
							// 처리되지 않은 건인 경우
							else
							{
								m_logger.LogDebug($"[Reject] MQ Message Received on [{{0}}] : {{1}}, Data = {{2}}", m_queueName, ea.RoutingKey, ea.Body.ToArray().GetString());

								// 처리 거부
								m_channel.BasicReject(ea.DeliveryTag, false);
							}
						}
					}
					catch (Exception ex)
					{
						NNException.Log(ex);
					}
				};

				// 이벤트 설정
				Consumer.Registered += OnConsumerRegistered;
				Consumer.Unregistered += OnConsumerUnregistered;
				Consumer.ConsumerCancelled += OnConsumerCancelled;
				Consumer.Shutdown += OnConsumerShutdown;

				m_channel.BasicConsume(m_queueName, false, Consumer);
			}
			catch (Exception ex)
			{
				NNException.Log(ex);
			}

			return Task.CompletedTask;
		}

		/// <summary>메시지 처리</summary>
		/// <param name="RoutingKey">라우팅 키</param>
		/// <param name="Body">내용</param>
		protected abstract Task<ResponseMqData> HandleMessage(string RoutingKey, byte[] Body);

		/// <summary>메세지 수신 등록 시 이벤트</summary>
		/// <param name="Sender">이벤트 발송 객체</param>
		/// <param name="e">이벤트 데이터</param>
		private void OnConsumerRegistered(object Sender, ConsumerEventArgs e)
		{
			m_logger.LogInformation($"[Rabbit MQ] Receiver is registered. {m_queueName}");
		}

		/// <summary>메세지 수신 해제 시 이벤트</summary>
		/// <param name="Sender">이벤트 발송 객체</param>
		/// <param name="e">이벤트 데이터</param>
		private void OnConsumerUnregistered(object Sender, ConsumerEventArgs e)
		{
			m_logger.LogInformation($"[Rabbit MQ] Receiver is unregistered. {m_queueName}");
		}

		/// <summary>메세지 수신 취소 시 이벤트</summary>
		/// <param name="Sender">이벤트 발송 객체</param>
		/// <param name="e">이벤트 데이터</param>
		private void OnConsumerCancelled(object Sender, ConsumerEventArgs e)
		{
			m_logger.LogInformation($"[Rabbit MQ] Receiver is cancelled. {m_queueName}");
		}

		/// <summary>메세지 수신 종료 시 이벤트</summary>
		/// <param name="Sender">이벤트 발송 객체</param>
		/// <param name="e">이벤트 데이터</param>
		private void OnConsumerShutdown(object Sender, ShutdownEventArgs e)
		{
			m_logger.LogInformation($"[Rabbit MQ] Receiver is shutdown. {m_queueName}");
		}
	}
}

File: core/common/mqmanage/RabbitMqReceiver.py
#!/usr/bin/env python3
"""
* Copyright (c) 2021 PSPACE, inc. KSAN Development Team ksan@pspace.co.kr
* KSAN is a suite of free software: you can redistribute it and/or modify it under the terms of
* the GNU General Public License as published by the Free Software Foundation, either version 
* 3 of the License.  See LICENSE for details
*
* 본 프로그램 및 관련 소스코드, 문서 등 모든 자료는 있는 그대로 제공이 됩니다.
* KSAN 프로젝트의 개발자 및 개발사는 이 프로그램을 사용한 결과에 따른 어떠한 책임도 지지 않습니다.
* KSAN 개발팀은 사전 공지, 허락, 동의 없이 KSAN 개발에 관련된 모든 결과물에 대한 LICENSE 방식을 변경 할 권리가 있습니다.
"""


import os, sys
if os.path.dirname(os.path.abspath(os.path.dirname(__file__))) not in sys.path:
    sys.path.append(os.path.dirname(os.path.abspath(os.path.dirname(__file__))))
#from service.service_mq_handle import MqServiceHandler
#import mqmanage
from mqmanage.RabbitMqConfiguration import RabbitMqConfiguration
from portal_api.apis import *

import json
import pika


class RabbitMqReceiver:
    """ 생성자
    Args:
        config: Rabbit MQ 설정 객체
    Returns:
        없음
    """

    def __init__(self, config: RabbitMqConfiguration, queueName: str, exchangeName: str, bindingKeys: tuple,
                 ServerId=None, ServiceList=None, LocalIpList=None, logger=None, MonConf=None, GlobalFlag=None):
        self.m_config = config
        self.m_queue_name = queueName
        self.m_exchange_name = exchangeName
        self.m_binding_keys = bindingKeys
        self.m_connection = None
        self.m_channel = None
        self.m_response = None
        self.m_correlation_id = None
        self.ServerId = ServerId
        self.ServiceList = ServiceList
        self.LocalIpList = LocalIpList
        self.logger = logger
        self.MonConf = MonConf
        self.GlobalFlag = GlobalFlag  # {'ServiceUpdated': Updated, 'DiskUpdated': Updated, 'NetworkUpdated': Checked}
        # Rabbit MQ 리스너 초기화
        while True:
            try:
                self.initializeRabbitMqListener(self.m_config)

                #result = self.m_channel.queue_declare(queue='', exclusive=True)
                #self.m_callback_queue = result.method.queue
                print(self.m_queue_name)
                self.m_channel.basic_consume(
                    queue=self.m_queue_name,
                    on_message_callback=self.on_response,
                    auto_ack=False)

                self.m_channel.start_consuming()
            except pika.exceptions.AMQPChannelError as err:
                logger.error('1rabbitmq server connection fail %s %d' % (str(err), sys.exc_info()[2].tb_lineno))
                time.sleep(5)
                continue
            except Exception as err:
                logger.error('2rabbitmq server connection fail %s %d' % (str(err), sys.exc_info()[2].tb_lineno))
                time.sleep(5)


    """ Rabbit MQ 리스너 초기화
    Args:
        config: Rabbit MQ 설정 객체
    Returns:
        없음
    """

    def initializeRabbitMqListener(self, config: RabbitMqConfiguration):

        # 계정 정보 생성
        credentials = pika.PlainCredentials(username=config.user, password=config.password)

        # 연결 파라미터 생성
        connectionParams = pika.ConnectionParameters(
            host=config.host
            , port=config.port
            , virtual_host=config.virtualHost
            , credentials=credentials
        )

        # 연결 객체 생성
        self.m_connection = pika.BlockingConnection(connectionParams)
        # 채널  생성
        self.m_channel = self.m_connection.channel()

        # 큐 설정
        mq_args = {}
        mq_args["x-queue-type"] = "quorum"
        mq_args["x-single-active-consumer"] = True

        self.m_channel.queue_declare(queue=self.m_queue_name,
                                     durable=True, exclusive=False, auto_delete=False, arguments=mq_args)
        # Exchange 설정
        self.m_channel.exchange_declare(exchange=self.m_exchange_name, exchange_type="topic")
        # 모든 바인딩 키 처리
        if len(self.m_binding_keys) > 0:
            for binding_key in self.m_binding_keys:
                self.m_channel.queue_bind(queue=self.m_queue_name, exchange=self.m_exchange_name,
                                          routing_key=binding_key)

    """ 소멸자
    Args:
        없음
    Returns:
        없음
    """

    def __del__(self):
        # 채널 연결이 되어 있는 경우, 채널 연결 해제
        if self.m_channel is not None:
            self.m_channel.close()
            self.m_channel = None

        # 연결이 되어 있는 경우, 연결 해제
        if self.m_connection is not None:
            self.m_connection.close()
            self.m_connection = None

    """ 메세지 수신 시 호출되는 콜백
    """

    def on_response(self, ch, method, props, body):
        #print(" [x] routing_key: %r, body: %r" % (method.routing_key, body))

        response = self.handleMessage(method.routing_key, body)
        ResponseReturn = self.MqReceiverHanlder(method.routing_key, body, response)

        if response.IsProcessed:
            # 연관 아이디가 있는 경우
            if props.correlation_id is not None:
                if len(props.correlation_id) > 0:
                    #response = json.dumps({"Result": "Success", "Code": "", "Messsage": ""})
                    #print(response)
                    ch.basic_publish(exchange='',
                                     routing_key=props.reply_to,
                                     properties=pika.BasicProperties(correlation_id=props.correlation_id),
                                     body=str(ResponseReturn))

            self.m_channel.basic_ack(delivery_tag=method.delivery_tag, multiple=False)  # in case job is processed, return
        else:
            self.m_channel.basic_reject(delivery_tag=method.delivery_tag, requeue=True)  # requeue

    def handleMessage(self, RoutingKey, Body) -> ResponseMqData:
        """
        Create ResponseMqData Format
        """
        response = ResponseMqData(EnumResponseResult.Error, "", "")
        #self.RabbitMqReceiver(RoutingKey, Body, response)
        return response

    def MqReceiverHanlder(self, RoutingKey, body, Response):
        """
        Message Queue Handler
        1. Disk
        2. Serivce
        3. Network
        """
        #print(RoutingKey, body)

        if RoutKeyDisk in RoutingKey or RoutKeyDiskRpcFinder.search(RoutingKey) or RoutKeyDiskPool in RoutingKey:
            ResponseReturn = MqDiskHandler(RoutingKey, body, Response, self.ServerId, self.GlobalFlag, self.logger)
            return ResponseReturn
            #Response.IsProcessed = True
        elif RoutKeyService in RoutingKey or RoutKeyServiceRpcFinder.search(RoutingKey):
            ResponseReturn = MqServiceHandler(self.MonConf, RoutingKey, body, Response, self.ServerId, self.ServiceList, self.LocalIpList, self.GlobalFlag,  self.logger)
            #Response.IsProcessed = True
            return ResponseReturn
        elif RoutKeyNetwork in RoutingKey or RoutKeyNetworkRpcFinder.search(RoutingKey):
            ResponseReturn = MqNetworkHandler(self.MonConf, RoutingKey, body, Response, self.ServerId, self.ServiceList, self.GlobalFlag, self.logger)
            Response.IsProcessed = True
            return ResponseReturn
        elif RoutKeyServerUpdate in RoutingKey or RoutKeyServerDel in RoutingKey or RoutKeyServerAdd in RoutingKey:
            ResponseReturn = MqServerHandler(self.MonConf, RoutingKey, body, Response, self.ServerId, self.logger)
            Response.IsProcessed = True
            return ResponseReturn
        else:
            self.logger.debug("Skip RouteKey: %s" % str(RoutingKey))
            Response.IsProcessed = True
            ResponseReturn = MqReturn(ResultSuccess)
            return ResponseReturn


#print(body)


"""
class Mq:
    def __init__(self, Host, Port, VirtualHost, User, Password, RoutingKey, ExchangeName, QueueName=None):
        self.config = RabbitMqConfiguration()
        self.config.host = Host
        self.config.port = Port
        self.config.virtualHost = VirtualHost
        self.config.user = User
        self.config.password = Password
        self.config.exchangeName = ExchangeName
        self.routingkey = RoutingKey
        self.queuename = QueueName

    def Sender(self, Data):
        mqSender = RabbitMqSender(self.config)
        mqSender.send(self.config.exchangeName, self.routingkey, Data)

    def Receiver(self):
        RabbitMqReceiver(self.config, self.queuename, self.config.exchangeName, bindingKeys=self.routingkey)

    def Rpc(self, Data):
        mqRpc = RabbitMqRpc(self.config)
        mqRpc.send(self.config.exchangeName, self.config.exchangeName, Data)
"""
